### PR TITLE
Initialize underlying data in Point

### PR DESCRIPTION
**Pull request by masterleinad**
Default initialization should be cleaner and the data should be private.

### CODE REVIEW

**Comment by masterleinad** [comment]
> Did you actually try on CUDA? I don't think this will work without C++17.

No, I only tried on my laptop, yet.

**Comment by masterleinad** [comment]
I guess this as close as it gets for having an aggregate-initializer like behavior and default value initialization.

File: src/details/ArborX_Point.hpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ class Point
   }
 
   KOKKOS_INLINE_FUNCTION
-  constexpr Point(Data data)
+  constexpr Point(Data const &data)
```
Comment (dalg24) [inline src/details/ArborX_Point.hpp]: I would have passed by value and moved into place.

Comment (dalg24) [reply src/details/ArborX_Point.hpp]: Anyway I see little value in the PR overall besides the couple places we had forgotten double braces to silent warnings

File: src/details/ArborX_Point.hpp [hunk 2]
```diff
@@ -14,32 +14,51 @@
 
 #include <Kokkos_Macros.hpp>
 
+#include <array>
```
Comment (dalg24) [inline src/details/ArborX_Point.hpp]: Not used anymore

File: src/details/ArborX_Point.hpp (hunk 1 above)
Comment (masterleinad) [reply src/details/ArborX_Point.hpp]: I see value in making sure that the underlying values are initialized, but won't fight for this pull request if no one else supports it.

Comment (dalg24) [reply src/details/ArborX_Point.hpp]: Then fix your constructor.  Your current implementation makes us pay for an extra copy if we pass a rvalue.

File: src/details/ArborX_Point.hpp [hunk 3]
```diff
@@ -14,32 +14,51 @@
...
+    double coords[3];
+  } _data;
+
 public:
   KOKKOS_INLINE_FUNCTION
-  constexpr double &operator[](unsigned int i) { return _coords[i]; }
+  constexpr Point()
+      : _data{}
+  {
+  }
```
Comment (dalg24) [inline src/details/ArborX_Point.hpp]: And my preference goes to a defaulted default constructor and an a default initializer when you declare the data member but at that point I am just nitpicking :)

File: src/details/ArborX_Point.hpp (hunk 1 above)
Comment (masterleinad) [reply src/details/ArborX_Point.hpp]: The assembly is the same (https://godbolt.org/z/ePlQOS) but I also like the value variant better.

File: src/details/ArborX_Point.hpp (hunk 3 above)
Comment (masterleinad) [reply src/details/ArborX_Point.hpp]: Couldn't get that to work in a `constexpr` context.

Comment (dalg24) [reply src/details/ArborX_Point.hpp]: Did you initialize `_data`?

Comment (masterleinad) [reply src/details/ArborX_Point.hpp]: If I define the default `Point` constructor as `default`, the compiler complains that `coords` is not initialized and I can't define a constructor for `Data` without disallowing aggregate-initialization we use in the other constructor.

Comment (dalg24) [reply src/details/ArborX_Point.hpp]: https://godbolt.org/z/bqDcYB

Comment (masterleinad) [reply src/details/ArborX_Point.hpp]: Yes, that works.

**Comment by aprokop** [comment]
I thought this PR was undecided?

**Comment by Rombur** [comment]
Where do you see that? Damien asked for a change in the constructor and the constructor has been changed

**Comment by aprokop** [comment]
I think @dalg24 had some reservations about merging this one in.

**Comment by Rombur** [comment]
Then he should have spelled it on the PR. How I am supposed to know.

**Pull request merged**